agent/src/beerocks/monitor/monitor_thread.cpp

```diff
@@ -494,7 +501,75 @@ void monitor_thread::after_select(bool timeout)
 #ifdef BEEROCKS_RDKB
         mon_rdkb_hal.process();
 #endif
+
+        /**
+         * If a Multi-AP Agent receives a Metric Reporting Policy TLV with AP Metrics Channel
+         * Utilization Reporting Threshold field set to a non-zero value for a given radio, it
+         * shall measure the channel utilization on that radio in each consecutive implementation-
+         * specific measurement period and, if the most recently measured channel utilization has
+         * crossed the reporting threshold in either direction (with respect to the previous
+         * measurement), it shall send an AP Metrics Response message to the Multi-AP Controller
+         * containing one AP Metrics TLV for each of the BSSs on that radio.
+         */
+        auto radio_node = mon_db.get_radio_node();
+        auto &info      = radio_node->ap_metrics_reporting_info();
+        if (0 != info.ap_channel_utilization_reporting_threshold) {
+            int elapsed_time_s =
+                std::chrono::duration_cast<std::chrono::seconds>(
+                    now - info.ap_metrics_channel_utilization_last_reporting_time_point)
+                    .count();
+
+            if (elapsed_time_s >= ap_metrics_channel_utilization_measurement_period_s) {
+                info.ap_metrics_channel_utilization_last_reporting_time_point = now;
+
+                on_channel_utilization_measurement_period_elapsed();
+            }
+        }
+    }
+}
+
+void monitor_thread::on_channel_utilization_measurement_period_elapsed()
+{
+    /**
+     * Measure current channel utilization on the radio.
+     */
+    uint8_t channel_utilization;
+    if (!mon_wlan_hal->get_channel_utilization(channel_utilization)) {
+        LOG(ERROR) << "Unable to get channel utilization";
+        return;
     }
+
+    /**
+     * Compare with previous measurement and if difference has crossed the reporting threshold,
+     * then send an AP Metrics Response message to the controller.
+     */
+    auto radio_node = mon_db.get_radio_node();
+    auto &info      = radio_node->ap_metrics_reporting_info();
+    if ((UINT8_MAX == info.ap_metrics_channel_utilization_reporting_value) ||
+        (abs(channel_utilization - info.ap_metrics_channel_utilization_reporting_value) >
+         info.ap_channel_utilization_reporting_threshold)) {
```

Comment:
This is not correct IMO. The threshold is not a difference like it is for RCPI - it's when you go over the threshold. "if the most recently measured channel utilization has crossed the
reporting threshold in either direction (with respect to the previous measurement)". So:
- the ap_metrics_channel_utilization_reporting_value should be stored every time, not only when the threshold is exceeded;
- we should check if current and previous are on the same side of the threshold.

The second can be done easily be checking the sign of (a - threshold) * (b - threshold).

So the code would look something like this:
```
    auto current_diff = channel_utilization - info.ap_channel_utilization_reporting_threshold;
    auto prev_diff = info.ap_metrics_channel_utilization_reporting_value - info.ap_channel_utilization_reporting_threshold;
    if (cur_diff * prev_diff < 0)
        ...
```

Please double-check the above w.r.t. types and overflows. It may need to but put in an int32_t to be sure. 

As for the initial case: since it's not specified, it doesn't matter on which side of the threshold we start. So I think initialising it to 0 is fine.

Reply:
Fixed